lib/Logger/LogBufferFeature.cpp

```diff
@@ -0,0 +1,201 @@
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Dr. Frank Celler
+////////////////////////////////////////////////////////////////////////////////
+
+#include "LogBufferFeature.h"
+
+#include "Basics/ArangoGlobalContext.h"
+#include "Basics/Mutex.h"
+#include "Basics/MutexLocker.h"
+#include "Basics/tri-strings.h"
+#include "Logger/LogAppender.h"
+#include "Logger/LoggerFeature.h"
+#include "Logger/Logger.h"
+#include "ProgramOptions/ProgramOptions.h"
+#include "ProgramOptions/Section.h"
+
+#ifdef _WIN32
+#include "Basics/win-utils.h"
+#endif
+
+#include <cstring>
+#include <utility>
+
+using namespace arangodb::basics;
+using namespace arangodb::options;
+
+namespace arangodb {
+
+LogBuffer::LogBuffer()
+    : _id(0), 
+      _level(LogLevel::DEFAULT), 
+      _topicId(0), 
+      _timestamp(0) {
+  memset(&_message[0], 0, sizeof(_message));
+}
+
+/// @brief logs to a fixed size ring buffer in memory
+class LogAppenderRingBuffer final : public LogAppender {
+ public:
+  LogAppenderRingBuffer() 
+      : LogAppender(),
+        _id(0) {
+    MUTEX_LOCKER(guard, _lock);
+    _buffer.resize(LogBufferFeature::BufferSize);
+  }
+
+ public:
+  void logMessage(LogMessage const& message) override {
+    auto timestamp = time(nullptr);
+
+    MUTEX_LOCKER(guard, _lock);
+
+    uint64_t n = _id++;
+    LogBuffer& ptr = _buffer[n % LogBufferFeature::BufferSize];
+
+    ptr._id = n;
+    ptr._level = message._level;
+    ptr._topicId = static_cast<uint32_t>(message._topicId);
+    ptr._timestamp = timestamp;
+    TRI_CopyString(ptr._message, message._message.c_str() + message._offset,
+                   sizeof(ptr._message) - 1);
+  }
+
+  std::string details() const override {
+    return std::string();
+  }
+
+  /// @brief return all buffered log entries
+  std::vector<LogBuffer> entries(LogLevel level, uint64_t start, bool upToLevel) {
+    std::vector<LogBuffer> result;
+    
+    uint64_t s = 0;
+    uint64_t n;
+
+    MUTEX_LOCKER(guard, _lock);
+
+    if (_id >= LogBufferFeature::BufferSize) {
+      s = _id % LogBufferFeature::BufferSize;
+      n = LogBufferFeature::BufferSize;
+    } else {
+      n = static_cast<uint64_t>(_id);
+    }
+
+    for (uint64_t i = s; 0 < n; --n) {
+      LogBuffer const& p = _buffer[i];
+
+      if (p._id >= start) {
+        if (upToLevel) {
+          if (static_cast<int>(p._level) <= static_cast<int>(level)) {
+            result.emplace_back(p);
+          }
+        } else {
+          if (p._level == level) {
+            result.emplace_back(p);
+          }
+        }
+      }
+
+      ++i;
+
+      if (i >= LogBufferFeature::BufferSize) {
+        i = 0;
+      }
+    }
+
+    return result;
+  }
+
+ private:
+  Mutex _lock;
+  uint64_t _id;
+  std::vector<LogBuffer> _buffer;
+};
+
+#ifdef _WIN32
+/// logs to the debug output windows in MSVC
+class LogAppenderDebugOutput final : public LogAppender {
```

Comment:
Yes, it may be confusing, but it still is correct.
This logger is invoking the `OutputDebugString` Win32-API function, which prints into the debug window.